File: code/seungwookim99/week3/boj_1149.java
// 백준 1149 : RGB거리
// Java 풀이
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    int N;
    ArrayList<House> house = new ArrayList<>();

    class House {
        int R,G,B;
        House (int R, int G, int B) {
            this.R = R;
            this.G = G;
            this.B = B;
        }
    }
    public void solution() throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in)); // for read
        StringBuilder sb = new StringBuilder(); // for output

        /* get input */
        N = Integer.parseInt(br.readLine());
        for (int i = 0 ; i < N ; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            house.add(new House(
                    Integer.parseInt(st.nextToken()),
                    Integer.parseInt(st.nextToken()),
                    Integer.parseInt(st.nextToken())
            ));
        }

        /* initialize */
        House dp[] = new House[N];
        for (int i = 0 ; i < N ; i++)
            dp[i] = new House(0, 0, 0);
        dp[0].R = house.get(0).R;
        dp[0].G = house.get(0).G;
        dp[0].B = house.get(0).B;

        /* dp */
        for (int i = 1 ; i < N ; i++) {
            dp[i].R = house.get(i).R + Math.min(dp[i-1].G, dp[i-1].B);
            dp[i].G = house.get(i).G + Math.min(dp[i-1].R, dp[i-1].B);
            dp[i].B = house.get(i).B + Math.min(dp[i-1].R, dp[i-1].G);
        }

        sb.append(Math.min(dp[N-1].R, Math.min(dp[N-1].G, dp[N-1].B)));
        System.out.println(sb);
    }

    public static void main(String[] args) throws Exception {
        new Main().solution();
    }
}

File: code/jonghyeonjo99/week3/boj_1149.py
n = int(input())

houses = []
dp = [[0 for _ in range(3)] for _ in range(n)]

for i in range(n):
  house = list(map(int,input().split()))
  houses.append(house)

dp[0][0] = houses[0][0]
dp[0][1] = houses[0][1]
dp[0][2] = houses[0][2]

for i in range(1,n):
  dp[i][0] = min(houses[i][0] + dp[i-1][1],houses[i][0] + dp[i-1][2])
  dp[i][1] = min(houses[i][1] + dp[i-1][0], houses[i][1] + dp[i-1][2])
  dp[i][2] = min(houses[i][2] + dp[i-1][0], houses[i][2] + dp[i-1][1])

result = min(dp[n-1])
print(result)

